core/server/worker/src/main/java/alluxio/worker/grpc/BlockReadHandler.java

```diff
@@ -462,28 +462,25 @@ private void runInternal() {
         }
         continue;
       }
-      if (error != null) {
-        try {
-          completeRequest(mContext);
-        } catch (Exception e) {
+      try {
+        completeRequest(mContext);
+      } catch (Exception e) {
+        if (error != null) {
           LOG.error("Failed to close the request.", e);
-        }
-        replyError(error);
-      } else if (eof || cancel) {
-        try {
-          completeRequest(mContext);
-        } catch (Exception e) {
-          LogUtils.warnWithException(LOG, "Exception occurred while completing read request, "
-                  + "EOF/CANCEL sessionId: {}. {}", mContext.getRequest().getSessionId(),
-              mContext.getRequest(), e);
-          setError(new Error(AlluxioStatusException.fromThrowable(e), true));
-        }
-        if (eof) {
-          replyEof();
         } else {
-          replyCancel();
+          LogUtils.warnWithException(LOG, "Exception occurred while completing read request, "
+                          + "EOF/CANCEL sessionId: {}. {}", mContext.getRequest().getSessionId(),
+                  mContext.getRequest(), e);
+          error = new Error(AlluxioStatusException.fromThrowable(e), true);
         }
       }
+      if (error != null) {
+        replyError(error);
+      } else if (eof) {
+        replyEof();
+      } else {
```

Comment:
do we need to check if cancel is needed here?

```suggestion
      } else if (cancel) {
```